### ISSUE TITLE
Update color palette accessibility checker

### ISSUE THREAD

**Issue by stephanieosan** [cfpb/design-system#1226]
**Which page is this about?** 
[Color](https://cfpb.github.io/design-system/foundation/color)

**What kind of issue is this?**
Content/tech issue

**Describe your issue** 
The 3rd-party color palette accessibility checker we link to is no longer live (link throws a 500 error). We need to find an alternative to this, ideally that can replicate the behavior of our previous palette contrast checker. There are many color contrast tools available online but few of them support pre-loading them with a color palette, a feature we need.

**Add any additional background info as needed** 

See the relevant [discussion](https://mattermost.local/cfpb/pl/a1og36ycyp8utjaada898zthgr).

Should be updated on the following pages: 
* [Color](https://cfpb.github.io/design-system/foundation/color)
* [Fonts](https://cfpb.github.io/design-system/foundation/fonts)

**Comment by contolini** [comment]
I loaded our colors into here: https://contrastchecker.com/?swatch_session=1612474035691

@anselmbradford @Scotchester Does the above link suffice?

**Comment by Scotchester** [comment]
> I loaded our colors into here: https://contrastchecker.com/?swatch_session=1612474035691
> 
> @anselmbradford @Scotchester Does the above link suffice?

Unfortunately, I don't think so. It only shows each color on white. But loading every possible color combination at once would be ridiculous. (Also, wow, the sample text is tiny.)

**Comment by nataliafitzgerald2** [comment]
I pulled the relevant comments from the Mattermost chat on this from April 2021. The next step would be to access these options and select the one that best suits our needs. 

Options @contolini posted: 
- some other options as well: https://github.com/oomphinc/colorcube#alternatives-other-great-tools-that-provided-inspiration

Options @Scotchester posted: 
- Not quite the same, but pretty cool: https://oomphinc.github.io/colorcube/. Can it auto-populate a color scheme? Hmm, maybe not as currently coded. Looks like it just submits via JavaScript without populating URL params.
- https://github.com/darekkay/a11y-contrast
- Looks like you could slooowly load our palette into the history of https://contrastchecker.com/ and then share that.
- https://onefrankguy.github.io/palette-a11y/ could also be good to fork/contribute.
- This is clunky, but functionally seems somewhat promising: http://www.daniel-soltis.com/palette-creator/. A different version that might be easier to use: http://www.daniel-soltis.com/contrast/ Only displays the color values in RGB, though, which is annoying.
- Not what we're looking for, but very cool: https://stylescratchpad.firebaseapp.com/
- Not perfect, but promising: https://accessiblecolors.app/ (GitHub: https://github.com/codewithkyle/accessible-colors)
- Not what we're looking for, but very cool: https://stylescratchpad.firebaseapp.com/
- And finally, a node package that we could use to power our own custom thing, if we wanted to go down that road: https://github.com/jameelmoses/is-contrast-checker

**Comment by sonnakim** [comment]
I didn't see the original tool, but is [this](https://contrast-grid.eightshapes.com/?version=1.1.0&background-colors=&foreground-colors=1e9642%0D%0A20aa3f%0D%0A66c368%0D%0Aaddc91%0D%0Ac7e5b3%0D%0Ae2efd8%0D%0A005e5d%0D%0A257675%0D%0A579695%0D%0A89b6b5%0D%0Ab4d2d1%0D%0Ad4e7e6%0D%0A0050b4%0D%0A0072ce%0D%0A4497dc%0D%0A7eb7e8%0D%0Aafd2f2%0D%0Ad6e8fa%0D%0A002d72%0D%0A254b87%0D%0A5674a3%0D%0A889cc0%0D%0Ab3c0d9%0D%0Ad3daeb%0D%0Aa01b68%0D%0Ab4267a%0D%0Ac55998%0D%0Ad486b2%0D%0Ae3b2cc%0D%0Af0d8e2%0D%0Ab63014%0D%0Ad14124%0D%0Add735d%0D%0Ae79e8e%0D%0Af0c3b8%0D%0Af7e0d9%0D%0Adc731c%0D%0Aff9e1b%0D%0Affb858%0D%0Affce8d%0D%0Affe1b9%0D%0Afff0dd%0D%0A745745%0D%0A8a6c57%0D%0Aa18573%0D%0Abaa496%0D%0Ad3c5bc%0D%0Ae7ddd7%0D%0A101820%0D%0A43484e%0D%0A5a5d61%0D%0A75787b%0D%0A919395%0D%0Ab4b5b6%0D%0Ad2d3d5%0D%0Ae7e8e9%0D%0Af7f8f9%0D%0Affffff&es-color-form__tile-size=compact&es-color-form__show-contrast=aaa&es-color-form__show-contrast=aa18&es-color-form__show-contrast=dnp) the kind of thing we're looking for? @nataliafitzgerald2 @contolini

**Comment by Scotchester** [comment]
> 
> 
> I didn't see the original tool, but is [this](https://contrast-grid.eightshapes.com/?version=1.1.0&background-colors=&foreground-colors=1e9642%0D%0A20aa3f%0D%0A66c368%0D%0Aaddc91%0D%0Ac7e5b3%0D%0Ae2efd8%0D%0A005e5d%0D%0A257675%0D%0A579695%0D%0A89b6b5%0D%0Ab4d2d1%0D%0Ad4e7e6%0D%0A0050b4%0D%0A0072ce%0D%0A4497dc%0D%0A7eb7e8%0D%0Aafd2f2%0D%0Ad6e8fa%0D%0A002d72%0D%0A254b87%0D%0A5674a3%0D%0A889cc0%0D%0Ab3c0d9%0D%0Ad3daeb%0D%0Aa01b68%0D%0Ab4267a%0D%0Ac55998%0D%0Ad486b2%0D%0Ae3b2cc%0D%0Af0d8e2%0D%0Ab63014%0D%0Ad14124%0D%0Add735d%0D%0Ae79e8e%0D%0Af0c3b8%0D%0Af7e0d9%0D%0Adc731c%0D%0Aff9e1b%0D%0Affb858%0D%0Affce8d%0D%0Affe1b9%0D%0Afff0dd%0D%0A745745%0D%0A8a6c57%0D%0Aa18573%0D%0Abaa496%0D%0Ad3c5bc%0D%0Ae7ddd7%0D%0A101820%0D%0A43484e%0D%0A5a5d61%0D%0A75787b%0D%0A919395%0D%0Ab4b5b6%0D%0Ad2d3d5%0D%0Ae7e8e9%0D%0Af7f8f9%0D%0Affffff&es-color-form__tile-size=compact&es-color-form__show-contrast=aaa&es-color-form__show-contrast=aa18&es-color-form__show-contrast=dnp) the kind of thing we're looking for? @nataliafitzgerald2 @contolini

❗ That looks awesome!!

**Comment by sonnakim** [comment]
Cool, thanks @scotchester! (We miss you!!)

**Comment by nataliafitzgerald2** [comment]
Hi @sonnakim 
I took a look at the tool and I have some thoughts. This tool does do a lot of what we would want but I wonder whether it provides so much information that the key information that a user would need to make a sound decision about their type would be lost. It is a pretty overwhelming output.  

**Some of the things I'd be looking for in our visualization:** 
- Contrast ratio for small versus large type (and that this information is easy to locate)
- Which combinations pass AA (large vs small type)
- A definition of what we consider to be small and large type. WCAG only includes type sizes in points but I think that in the previous tool (or somewhere else) @Scotchester had documented a pixel translation. We'll have to track that down. 
- A list that is simple and not overwhelming that helps a less fluent user to make type color decisions that are both compliant and on brand. 
- I don't have a screenshot of the tool that @Scotchester built but I think that everything appeared on the screen at once without the need for horizontal scrolling. Do we have any screen shots of the original tool for reference? 
 
I think it would make sense to simplify the output as much as possible in order to make it most applicable to the work. So I wonder whether a tool that includes only the most realistic combinations would be more helpful.  
- In the page content within the Design System we include "Fully accessible combinations," "Non-accessible combinations," and "Undesirable combinations." I imagine that as an exercise we could sort all of the combinations into these 3 categories and then decide which information is worth visualizing for the user/most helpful.
- Our primary goal is to make type easy to read for our users. So all type color decisions should be made with this in mind.
- In the majority of cases of type in our work we stick with white, gray, or black type. 
   - One exception to this is that we use some type colors for interactive elements. We use Pacific for links/buttons, Teal for visited links, Navy for active links/button, Red for destructive links/buttons but in these cases we always use white text on the colored background or colored text on a white background. 
- It's helpful to show green type on white and white type on green because this is our brand color and users will be interested in whether this is contrast compliant for large and small type. 
- In general, we don't place type of one color on a background of a different color - even if there is sufficient contrast (for example teal type on a gold background). So it may make sense to not show these types of combinations. This would reduce the number of combinations by a lot. 
- For instances where we have a colored background, most notably hero graphics, social media graphics, featured content modules, and videos the best practice is to place white type on a dark background or black text on a light background (ensuring that the combinations have the proper contrast ratio).
- It is not brand compliant to place green on black or black on green, even when there is sufficient contrast. 

Maybe as a first step (MVP) we can use the tool that @sonnakim has populated so that in the near term we don't continue to have a dead link on the Design System page. And, as a next step we can try to develop something that is more specific to our goal of providing users with a tool for selecting type colors that have the proper contrast ratio AND are aligned with our brand guidelines and best practices.

**Comment by sonnakim** [comment]
@nataliafitzgerald I wonder if there's a limited number of type/background color combinations that are both brand complaint and accessible, and if so, perhaps we add them directly to the color page in the Design System. So, the next step would be to see how many combinations we're actually talking about. It may not be worth setting up a third party tool if it's easy enough to add these combinations directly to the Design System.

**Comment by caheberer** [comment]
I'm curious if there's a more simplified solution. Personally, I never used the previous tool because I knew the general rule that black text on any of our colors at 20% and white text on any of our 100%+ saturation colors (except gold) is accessible. If I wanted to use a different color or tint I would check that combination individually on the WebAIM tool. It rarely comes up because as Natalia mentions above we don't recommend using colored type on a colored background, or even colored type much at all. Most colored type that we use are embedded in our patterns.

Would it be easier to provide general rules on the design system like this?

**Comment by caheberer** [comment]
During scrum we determined I would move forward with this proposed solution. Below is a mock-up of how I think the Fonts page could be updated to include more simplified guidance around accessible color combinations.

While we are at it, what do we think of changing the title of the page from "Fonts" to "Type"?

@sonnakim @designlanguage @ajbush mind letting me know what you think?

This would follow "Spacing" within the "Guidelines" section of the [Fonts page](https://cfpb.github.io/design-system/foundation/fonts).
![Color page mockup](https://user-images.githubusercontent.com/4739887/145241125-e94b9846-69c7-42b6-8bb7-3a29db24ce1d.png)

**Comment by sonnakim** [comment]
@caheberer Nice! I like the simplicity and focus of what you've done here. It might be worth running this by one of the production designers as well to get their thoughts. Just a few small comments: 

- I like the idea of renaming the page to type — it does seem more appropriate. If you do rename it, I think we should add a redirect from the existing URL, as folks may be linking to this page from elsewhere. I think @brandon-rusk has handled redirects before, so you can let him know if you do rename it. I believe the nav will also have to be updated to link to the renamed page — that's something that you should have edit rights to do, but let me know if you run into issues.

---

![image](https://user-images.githubusercontent.com/9451445/145457562-5832c0f3-744e-479b-8a99-2bca79b7895e.png)

- Should we add the sentence "the only exception to colored type falls within illustration or other graphics" to the "Colored Type" section, and then add a jump link? Or combine these into one section? Otherwise it could sound like a contradiction. If there's a common enough use case for adding colored type in graphics, maybe we should make that clear, so there's no confusion. 

---

![image](https://user-images.githubusercontent.com/9451445/145457324-c5606c86-bc48-45ab-9484-7e9b9169b467.png)

I think "saturation of 100% and above" is a typo? 

---

- Should this say "Web Content Accessibility Guideline (WCAG) standards "? It's unusual to see "WCA" as a standalone acronym

**Comment by caheberer** [comment]
Thanks for the feedback, @sonnakim  I'll work on incorporating your suggestions.

**Comment by caheberer** [comment]
Here's the revised version. @sonnakim let me know if you have any additional feedback. I'm going to ping additional designers/production designers to take a look as well.

![Color page mockup](https://user-images.githubusercontent.com/4739887/145629164-f36030bb-80a8-4ea1-b955-fdc3eb643529.png)
.

**Comment by sonnakim** [comment]
@caheberer looks great!

**Comment by caheberer** [comment]
AJ reviewed and liked the simplicity. He also suggested incorporating the inclusion of "A few colors allow for slightly darker saturations while still maintaining accessibility." To cover the instances where we use darker saturations with black type, like in heroes. 

I made the updates and published the page.
